### PR TITLE
[fix] Make operations on `individualDeletedMessages` in lock scope

### PR DESCRIPTION
**Pull request by dao-jun**
### Motivation

In https://github.com/apache/pulsar/pull/22908 we introduced `ConcurrentRoaringBitSet` which is based on `StampLock` and `RoaringBitmap` to optimize the memory usage and GC pause on `BitSet`.

However, there is a concurrency issue on `ConcurrentRoaringBitSet`.

It will throw NPE when calling `ConcurrentRoaringBitSet#get` and `ConcurrentRoaringBitSet#set` in multiple threads, the situation is a little similar with  https://github.com/apache/pulsar/issues/18388. 

see: 
[RoaringBitmap#add](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java#L1152-L1162)
[RoaringBitmap#get](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java#L1683-L1691)

It will throw NPE if use StampLock, the situation is a little similar with https://github.com/apache/pulsar/issues/18388

### Modifications

1. Remove `ConcurrentBitSet`
2. Rename `ConcurrentOpenLongPairRangeSet` to `OpenLongPairRangeSet` and mark it as NotThreadSafe.
3. Make all the operations on `ManageCursorImpl#individualDeletedMessages` in ReadWriteLock scope.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Review by lhotari — CHANGES_REQUESTED** [review]
I think that we need to find another solution. ReadWriteLock adds a lot more overhead than StampedLock.

I wonder if it would be a viable option to catch exceptions and retry with a read lock if that happens?

**Comment by dao-jun** [comment]
> I think that we need to find another solution. ReadWriteLock adds a lot more overhead than StampedLock.

Yes, but RoaringBitmap is not designed for Concurrency at all, and the PR is a quick fix, we can make further improvements in the future.

**Comment by lhotari** [comment]
> > I wonder if it would be a viable option to catch exceptions and retry with a read lock if that happens?
> 
> Then we may catch a lot of exceptions when a broker is in a large throughput, I'm not sure if the cost is less than RWLock or not.

That's a valid concern, we should investigate the different choices and experiment.

**Comment by lhotari** [comment]
I think that we should revert the migration to RoaringBitSet in branch-3.0, branch-3.2 and branch-3.3 so that we don't need to rush with the solution.

**Comment by lhotari** [comment]
I reverted the changes in branch-3.0, branch-3.2 and branch-3.3. Here's the PR to revert the change in master branch: #22968 . It's better to have a fresh start with a proper fix that is validated so that it doesn't cause performance regressions and also addresses the concurrency issues. The concern about switching to ReadWriteLock is about it causing a performance regression. It's possible that it's not a valid concern, but let's validate that before applying the solution.

**Comment by dao-jun** [comment]
I did a less rigorous test:
```java
    @Test
    public void test() {
        long start = System.currentTimeMillis();
        CountDownLatch latch = new CountDownLatch(2);
        ConcurrentRoaringBitSet bitSet = new ConcurrentRoaringBitSet();
        new Thread(() -> {
            for (int i = 0; i < 100000000; i++) {
                bitSet.set(1);
            }
            latch.countDown();
        }).start();
        new Thread(() -> {
            for (int i = 0; i < 100000000; i++) {
                bitSet.get(1);
            }
            latch.countDown();
        }).start();

        try {
            latch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("Time: " + (System.currentTimeMillis() - start));
    }
```

I started 2 threads to call get/set methods on `ReadWriteLock`/`StampLock` based ConcurrentRoaringBitSet, each thread looping 100 million times.
For `ReadWriteLock` based `ConcurrentRoaringBitSet`, the total durations are around 9.5s
For `StampLock` base `ConcurrentRoaringBitSet`, the total durations are around 8.5s.

Maybe we don't need to worry about the performance regression?

**Comment by dao-jun** [comment]
When we do Readonly operations on StampLock based ConcurrentRoaringBitSet, it does faster than ReadWriteLock(about 5 times faster), but in the case we use `ConcurrentRoaringBitSet` is Read and Write(about 1:1).

**Comment by lhotari** [comment]
> When we do Readonly operations on StampLock based ConcurrentRoaringBitSet, it does faster than ReadWriteLock(about 5 times faster), but in the case we use `ConcurrentRoaringBitSet` is Read and Write(about 1:1).

In Pulsar we have https://github.com/apache/pulsar/tree/master/microbench module with JMH. I think JMH is better for comparisons. For Pulsar, the efficiency also matters so the comparison might not be that simple.

btw. In Pulsar `ConcurrentOpenLongPairRangeSet` is only used in `RangeSetWrapper` and the only usage of that is in `ManagedCursorImpl` for `individualDeletedMessages`. In many cases, the operations on `individualDeletedMessages` are already protected by the `ReadWriteLock` field `lock` in `ManagedCursorImpl`. 
It might be better to make the lock usage consistent. We wouldn't need `ConcurrentRoaringBitSet` in the Pulsar code base in that case as long as we document that `ConcurrentOpenLongPairRangeSet` isn't really thread safe. The thread safe solution could use the old solution.

**Comment by dao-jun** [comment]
> btw. In Pulsar ConcurrentOpenLongPairRangeSet is only used in RangeSetWrapper and the only usage of that is in ManagedCursorImpl for individualDeletedMessages. In many cases, the operations on individualDeletedMessages are already protected by the ReadWriteLock field lock in ManagedCursorImpl.
It might be better to make the lock usage consistent. We wouldn't need ConcurrentRoaringBitSet in the Pulsar code base in that case as long as we document that ConcurrentOpenLongPairRangeSet isn't really thread safe. The thread safe solution could use the old solution.

It makes sense, I addressed this, PTAL

**Comment by lhotari** [comment]
> It makes sense, I addressed this, PTAL

@dao-jun Looks good, I'll soon review in more detail. Please update the PR title and description so that it describes the motivation and modifications of this PR more accurately.

**Review by lhotari — CHANGES_REQUESTED** [review]
Please use write lock for `individualDeletedMessages.resetDirtyKeys();` call in `buildIndividualDeletedMessageRanges` method.

**Comment by lhotari** [comment]
Since the previous change #22908 was rollbacked by #22968, please rebase the changes.

**Review by lhotari — COMMENTED** [review]
> Rename ConcurrentOpenLongPairRangeSet to OpenLongPairRangeSet and mark it as NotThreadSafe.

I guess this change and the switch to use RoaringBitSet (in version 1.1.0) was lost in rebasing?

**Comment by lhotari** [comment]
> Please use write lock for `individualDeletedMessages.resetDirtyKeys();` call in `buildIndividualDeletedMessageRanges` method.

This is actually a real bug in the current implementation and needs to be fixed even if we wouldn't switch to use RoaringBitMap's RoaringBitSet.

**Comment by lhotari** [comment]
> > Rename ConcurrentOpenLongPairRangeSet to OpenLongPairRangeSet and mark it as NotThreadSafe.
> 
> I guess this change and the switch to use RoaringBitSet (in version 1.1.0) was lost in rebasing?

One possibility would be to complete this PR by switching to the non-thread version of ConcurrentOpenLongPairRangeSet which using ordinary BitSet in this PR and then switch to use RoaringBitSet in a follow up PR.

**Comment by dao-jun** [comment]
> > 将 ConcurrentOpenLongPairRangeSet 重命名为 OpenLongPairRangeSet 并将其标记为 NotThreadSafe。
> 
> 我猜测这个改变和使用 RoaringBitSet（在版本 1.1.0 中）的切换在重新定基时丢失了？

Yes, after the PR merged, I'll switch BitSet to RoaringBitSet again.

**Comment by dao-jun** [comment]
> > Please use write lock for `individualDeletedMessages.resetDirtyKeys();` call in `buildIndividualDeletedMessageRanges` method.
> 
> This is actually a real bug in the current implementation and needs to be fixed even if we wouldn't switch to use RoaringBitMap's RoaringBitSet.

Yes, `individualDeletedMessages.resetDirtyKeys()` is a WRITE operation, but it just requires a READ lock.

**Review by lhotari — APPROVED** [review]

**Comment by lhotari** [comment]
LGTM, good work @dao-jun

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/OpenLongPairRangeSet.java [hunk 1]
```diff
@@ -54,15 +56,15 @@ public class ConcurrentOpenLongPairRangeSet<T extends Comparable<T>> implements
     private volatile boolean updatedAfterCachedForSize = true;
     private volatile boolean updatedAfterCachedForToString = true;
 
-    public ConcurrentOpenLongPairRangeSet(LongPairConsumer<T> consumer) {
+    public OpenLongPairRangeSet(LongPairConsumer<T> consumer) {
         this(1024, true, consumer);
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/OpenLongPairRangeSet.java:60]: Just wondering if thread safe option should be removed or default to false?

Comment (dao-jun) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/OpenLongPairRangeSet.java:60]: Maybe in the followup PR(RoaringBitSet)?

**Pull request merged**